src/ui/CategoryFacet/CategoryFacetDebug.ts

```diff
@@ -0,0 +1,90 @@
+import { CategoryFacet } from './CategoryFacet';
+import { IBuildingQueryEventArgs, QueryEvents, IQuerySuccessEventArgs } from '../../events/QueryEvents';
+import { IGroupByResult } from '../../rest/GroupByResult';
+import { QueryBuilder } from '../Base/QueryBuilder';
+import { sortBy, chain } from 'underscore';
+import { Logger } from '../../misc/Logger';
+
+class PathMap {
+  get: (key: string) => PathMap;
+  set: (key: string, value?: PathMap) => void;
+}
+
+export class CategoryFacetDebug {
+  private positionInQuery: [number, number];
+  constructor(private categoryFacet: CategoryFacet) {
+    this.categoryFacet.bind.onRootElement<IBuildingQueryEventArgs>(QueryEvents.buildingQuery, args => this.addGroupBy(args));
+    this.categoryFacet.bind.onRootElement<IQuerySuccessEventArgs>(QueryEvents.querySuccess, args => {
+      args.results.groupByResults
+        .slice(this.positionInQuery[0], this.positionInQuery[1])
+        .forEach(groupByResult => CategoryFacetDebug.analyzeResults(groupByResult, this.categoryFacet.options.delimitingCharacter));
+    });
+  }
+
+  private addGroupBy(args: IBuildingQueryEventArgs) {
+    const firstPositionInQuery = args.queryBuilder.groupByRequests.length;
+    if (this.categoryFacet.activePath.length == 0) {
+      this.positionInQuery = [firstPositionInQuery, firstPositionInQuery + 1];
+      this.addGroupByForEmptyPath(args.queryBuilder);
+    } else {
+      const path = this.categoryFacet.activePath;
+      this.positionInQuery = [firstPositionInQuery, firstPositionInQuery + path.length];
+      this.addGroupByForEachPathElement(args.queryBuilder, path);
+    }
+  }
+
+  private addGroupByForEmptyPath(queryBuilder: QueryBuilder) {
+    queryBuilder.groupByRequests.push({
+      field: this.categoryFacet.options.field as string,
+      injectionDepth: this.categoryFacet.options.injectionDepth
+    });
+  }
+
+  private addGroupByForEachPathElement(queryBuilder: QueryBuilder, path: string[]) {
+    path.forEach(pathElement => {
+      this.categoryFacet.categoryFacetQueryController.addDebugGroupBy(queryBuilder, pathElement);
+    });
+  }
+
+  /**
+   *  This method expects group by results from a hierarchical field. It will find values with missing parents and values that are
```

Comment:
Since the class is not documented, the method documentation will not be outputted in the reference documentation. This is fine because it is already exposed in the CategoryFacet component.

Reply:
Are you sure? I've seen methods being documented alone without their class

Reply:
Yes. The FacetQueryController has its methods documented without having the class documented and it does not show up in the reference documentation. Anyway I removed the documentation for this method.